### PR TITLE
Experiment reducing yields

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/xproc/MethodYield.java [hunk 1]
```diff
@@ -56,7 +56,7 @@
   final ExplodedGraph.Node node;
   private final Map<String, Map<String, Set<Flow>>> cachedFlows = new HashMap<>();
   private final MethodBehavior behavior;
-  List<ConstraintsByDomain> parametersConstraints;
+  public List<ConstraintsByDomain> parametersConstraints;
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/xproc/MethodYield.java]: MINOR Make parametersConstraints a static final constant or non-public and provide accessors if needed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AClassVariableVisibilityCheck)

File: java-frontend/src/main/java/org/sonar/java/se/xproc/MethodBehavior.java [hunk 2]
```diff
@@ -90,13 +103,32 @@ public void createYield(ExplodedGraph.Node node, boolean storeNodeForReporting)
...
+        org.objectweb.asm.Type[] argumentTypes = org.objectweb.asm.Type.getArgumentTypes(signature.substring(signature.indexOf('(')));
+        constraints = cleanup(constraints, argumentTypes[index]);
       }
       yield.parametersConstraints.add(constraints);
+      index++;
+    }
+  }
+
+  private static ConstraintsByDomain cleanup(ConstraintsByDomain constraints, org.objectweb.asm.Type argumentType) {
+    if (constraints == null || constraints.isEmpty()) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/xproc/MethodBehavior.java]: MAJOR Remove this expression which always evaluates to "false" [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2589)

File: java-frontend/src/main/java/org/sonar/java/se/xproc/MethodBehavior.java [hunk 3]
```diff
@@ -153,6 +185,58 @@ public boolean isComplete() {
   public void completed() {
     this.complete = true;
     this.visited = true;
+    reduceYields();
+  }
+
+  private void reduceYields() {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/xproc/MethodBehavior.java]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 28 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [MethodBehavior.java#L69](https://github.com/SonarSource/sonar-java/blob/2e79fd1b031a61e38775af17600366eedda9238d/java-frontend/src/main/java/org/sonar/java/se/xproc/MethodBehavior.java#L69): Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [BytecodeEGWalker.java#L772](https://github.com/SonarSource/sonar-java/blob/2e79fd1b031a61e38775af17600366eedda9238d/java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java#L772): Take the required action to fix the issue indicated by this comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [MethodBehavior.java#L69](https://github.com/SonarSource/sonar-java/blob/35cbd3c09b333704189262d83bef86799571faf1/java-frontend/src/main/java/org/sonar/java/se/xproc/MethodBehavior.java#L69): Refactor this method to reduce its Cognitive Complexity from 16 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [BytecodeEGWalker.java#L772](https://github.com/SonarSource/sonar-java/blob/35cbd3c09b333704189262d83bef86799571faf1/java-frontend/src/main/java/org/sonar/java/bytecode/se/BytecodeEGWalker.java#L772): Take the required action to fix the issue indicated by this comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1134)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by m-g-sonar** [comment]
Closing the PR because some of this has been already introduced into production code in context of [SONARJAVA-2584](https://jira.sonarsource.com/browse/SONARJAVA-2584).

**Pull request closed without merging**